### PR TITLE
Documentation and running visualizer under uwsgi.

### PR DESCRIPTION
**Pull request by b1-88er**
- [x] @io41

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,82 @@
 [![CircleCI](https://circleci.com/gh/fieldaware/liveprofiler.svg?style=svg)](https://circleci.com/gh/fieldaware/liveprofiler)
 [![PyPI version](https://badge.fury.io/py/liveprofiler.svg)](https://badge.fury.io/py/liveprofiler)
 
-# live-profiler
-Middleware for profiling live services
+# Liveprofiler visualizer and collector
+
+This is a service that collects and presents samples generated by [livepofier_sampler](https://github.com/fieldaware/liveprofiler_sampler).
+It is fallow up work based on https://github.com/nylas/nylas-perftools. So most of the credit goes to: nylas.com
```
Comment (io41) [inline README.md]: typo `fallow` -> `follow`

File: example/uwsgi.ini [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[uwsgi]
+http-socket = 0.0.0.0:9999
+wsgi-file = liveprofiler/wsgi.py
+master = true
+enable-threads = true
+processes = 1
+threads = 2
+app_config=example/config.ini
+honour-stdin = true
```
Comment (io41) [inline example/uwsgi.ini]: this shouldn't be here I take it?

File: example/uwsgi.ini [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+# enable child processes running cpython to trap OS signals
+py-call-osafterfork = true
```
Comment (io41) [inline example/uwsgi.ini]: This for the sampler?

File: liveprofiler/collector.py [hunk 4]
```diff
@@ -30,19 +29,26 @@ def fetch_samples(host):
...
+
+def _collect(dbpath, hosts, secret_header):
+    db = model.ProflingModel(dbpath)
     collected = 0
-    for host in current_app.config['collector']['hosts']:
-        samples = fetch_samples(host)
+    for host in hosts:
+        samples = fetch_samples(host, secret_header)
         if not samples:
             continue
```
Comment (io41) [inline liveprofiler/collector.py]: Probably worth adding a debug log statement here?

File: example/uwsgi.ini (hunk 3 above)
Comment (eddwardo) [inline example/uwsgi.ini]: yes, I'll add the comment

File: example/uwsgi.ini (hunk 2 above)
Comment (eddwardo) [inline example/uwsgi.ini]: to delete

File: liveprofiler/collector.py (hunk 4 above)
Comment (eddwardo) [inline liveprofiler/collector.py]: it should be warning actually.

File: liveprofiler/wsgi.py [hunk 5]
```diff
@@ -4,6 +4,16 @@
...
 logging.config.fileConfig(CONFIG)
+log = logging.getLogger(__name__)
 
 from liveprofiler import app
 application = app.make_app(CONFIG)
+
+from liveprofiler import collector
+import uwsgidecorators
+
+@uwsgidecorators.timer(2)
```
Comment (io41) [inline liveprofiler/wsgi.py]: Can we move the interval to the config? Also, a comment on the interval units would be good. Any reason for the above value?

If not, I'd suggest 1 minute as a sensible default.

File: liveprofiler/app.py [hunk 6]
```diff
@@ -33,7 +46,7 @@ def run(cfg_path, port, debug):
     app = make_app(cfg_path)
     import logging.config
     logging.config.fileConfig(cfg_path)
-    app.run(host='0.0.0.0', port=port, debug=debug)
+    app.run(host='0.0.0.0', port=port, debug=debug, threaded=True)
```
Comment (io41) [inline liveprofiler/app.py]: host should probably be configurable, but with the current default of `0.0.0.0`

File: liveprofiler/collector.py [hunk 7]
```diff
@@ -36,6 +36,7 @@ def _collect(dbpath, hosts, secret_header):
     for host in hosts:
         samples = fetch_samples(host, secret_header)
         if not samples:
+            log.warning('No samples has been collected from: {}'.format(host))
```
Comment (io41) [inline liveprofiler/collector.py]: typo `has` -> `have`

File: liveprofiler/app.py (hunk 6 above)
Comment (eddwardo) [inline liveprofiler/app.py]: it is used only for running it locally.

File: README.md [hunk 8]
```diff
@@ -1,5 +1,82 @@
...
+> uwsgi example/uwsgi.ini
+```
+
+Then visit `http://localhost:9999/`.
+
+# uWSGI integration
+
+* Use at least 2 workers (processes) to run visualizer - one for visulizer and one for collector
+* Visualizer uses [uWSGI decorators](http://uwsgi-docs.readthedocs.io/en/latest/PythonDecorators.html) to sechdule collector calls - therefore `master = true` has to be enabled in uWSGI configuration
+* `app_config` is a path to the visualizer configuration file - it has to be set and it should be absolute path
```
Comment (io41) [inline README.md]: What if it's not an absolute path? In the example directory's uwsgi config file it's a relative path...

Comment (eddwardo) [inline README.md]: if path is not absolute, it will relative to $PWD where uwsgi is being run.

File: liveprofiler/app.py (hunk 6 above)
Comment (io41) [inline liveprofiler/app.py]: 👍

File: README.md (hunk 8 above)
Comment (io41) [inline README.md]: `it has to be set and it should be absolute path` -> `it must be set and it recommend it be an absolute path`

File: liveprofiler/wsgi.py [hunk 9]
```diff
@@ -12,8 +12,16 @@
...
-    log.info('Crontick from uwsgi timer, sig: {}'.format(sig))
+def collect():
     cfg = app.get_config(CONFIG)
-    collector._collect(cfg['global']['dbpath'], cfg['collector']['hosts'], cfg['collector']['secret_header'])
+    interval = int(cfg['collector'].get('interval', 60))
+    dbpath = cfg['global']['dbpath']
+    hosts = cfg['collector']['hosts']
+    secret_header = cfg['collector']['secret_header']
+
+    @uwsgidecorators.timer(interval)
```
Comment (io41) [inline liveprofiler/wsgi.py]: Will this not start a new timer every time collect() is called?

File: example/config.ini [hunk 10]
```diff
@@ -2,6 +2,7 @@
 dbpath = /tmp/liveprofiler/
 
 [collector]
+interval = 2
```
Comment (io41) [inline example/config.ini]: I'd like to see a comment here saying what units these are

Comment (io41) [inline example/config.ini]: Assuming seconds, a comment should probably be added here saying that for production 60s is recommended.

File: liveprofiler/wsgi.py (hunk 9 above)
Comment (io41) [inline liveprofiler/wsgi.py]: I would have guessed you need to a postfork hook in which to register the timer...

Comment (eddwardo) [inline liveprofiler/wsgi.py]: yes, but it is called only once - at the bottom of the module

Comment (io41) [inline liveprofiler/wsgi.py]: Ah yes, so `collect()` doesn't actually "collect", it starts the collection process. Suggest renaming it to `start_collector` or something similar that reflects what it's actually doing.

Comment (eddwardo) [inline liveprofiler/wsgi.py]: updated.

**Pull request merged**